Panic while running daemon

<!-- Output From `ipfs version --all`) -->
#### Version information: 
master

<!-- Bug, Feature, Enhancement, Etc -->
#### Type:
Panic

<!-- from `P0` "Critical" to `P5` "Relatively Unimportant") -->
#### Priority:
P1

#### Description:
Got a panic while running the daemon.

```
Initializing daemon...
Swarm listening on /ip4/10.5.8.247/tcp/4001
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip6/::1/tcp/4001
API server listening on /ip4/127.0.0.1/tcp/5001
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
Daemon is ready
panic: runtime error: index out of range

goroutine 95 [running]:
panic(0xc00500, 0xc420010110)
        /home/whyrusleeping/go/src/runtime/panic.go:500 +0x1a1
github.com/ipfs/go-ipfs/merkledag.(*Node).getPBNode(0xc42502c4b0, 0xc42023ec88)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/merkledag/coding.go:61 +0x33f
github.com/ipfs/go-ipfs/merkledag.(*Node).Marshal(0xc42502c4b0, 0xc422c3e760, 0xc422c3e760, 0x11387e0, 0xc422c3e760, 0x2)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/merkledag/coding.go:45 +0x2f
github.com/ipfs/go-ipfs/merkledag.(*Node).EncodeProtobuf(0xc42502c4b0, 0x0, 0x3, 0x22, 0x22, 0x22, 0x30)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/merkledag/coding.go:80 +0xb3
github.com/ipfs/go-ipfs/merkledag.(*Node).Multihash(0xc42502c4b0, 0xc42023ed98, 0x449915, 0xc42023eda8)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/merkledag/node.go:271 +0x34
github.com/ipfs/go-ipfs/merkledag.(*Node).Key(0xc42502c4b0, 0xc420063020, 0x0)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/merkledag/node.go:249 +0x2b
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).provideCollector(0xc420234a00, 0x113ae20, 0xc420130d80)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:138 +0x2c4
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).startWorkers.func4(0x113f840, 0xc42019cde0)
        /home/whyrusleeping/gopkg/src/github.com/ipfs/go-ipfs/exchange/bitswap/workers.go:40 +0x3c
gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go.func1(0xc4208fe980, 0xc42019cde0, 0xc42019cf00)
        /home/whyrusleeping/gopkg/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:112 +0x3c
created by gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go
        /home/whyrusleeping/gopkg/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:115 +0x375
```






<!--
This is for you! Please read, and then delete this text before posting it.
The go-ipfs issues are only for bug reports and directly actionable features.
Read https://github.com/ipfs/community/blob/master/contributing.md#reporting-issues if your issue doesn't fit either of those categories.
Read https://github.com/ipfs/go-ipfs/blob/master/docs/github-issue-guide.md if you are not sure how to fill in this issue.
-->

--- a/exchange/bitswap/bitswap.go
+++ b/exchange/bitswap/bitswap.go
@@ -90,7 +90,7 @@ func New(parent context.Context, p peer.ID, network bsnet.BitSwapNetwork,
 		network:       network,
 		findKeys:      make(chan *blockRequest, sizeBatchRequestChan),
 		process:       px,
-		newBlocks:     make(chan blocks.Block, HasBlockBufferSize),
+		newBlocks:     make(chan key.Key, HasBlockBufferSize),
 		provideKeys:   make(chan key.Key, provideKeysBufferSize),
 		wm:            NewWantManager(ctx, network),
 	}
@@ -137,7 +137,7 @@ type Bitswap struct {
 
 	process process.Process
 
-	newBlocks chan blocks.Block
+	newBlocks chan key.Key
 
 	provideKeys chan key.Key
 
@@ -308,12 +308,17 @@ func (bs *Bitswap) HasBlock(blk blocks.Block) error {
 		return err
 	}
 
+	// NOTE: There exists the possiblity for a race condition here.  If a user
+	// creates a node, then adds it to the dagservice while another goroutine
+	// is waiting on a GetBlock for that object, they will receive a reference
+	// to the same node. We should address this soon, but i'm not going to do
+	// it now as it requires more thought and isnt causing immediate problems.
 	bs.notifications.Publish(blk)
 
 	bs.engine.AddBlock(blk)
 
 	select {
-	case bs.newBlocks <- blk:
+	case bs.newBlocks <- blk.Key():
 		// send block off to be reprovided
 	case <-bs.process.Closing():
 		return bs.process.Close()

--- a/exchange/bitswap/workers.go
+++ b/exchange/bitswap/workers.go
@@ -127,17 +127,17 @@ func (bs *Bitswap) provideCollector(ctx context.Context) {
 
 	for {
 		select {
-		case blk, ok := <-bs.newBlocks:
+		case blkey, ok := <-bs.newBlocks:
 			if !ok {
 				log.Debug("newBlocks channel closed")
 				return
 			}
 
 			if keysOut == nil {
-				nextKey = blk.Key()
+				nextKey = blkey
 				keysOut = bs.provideKeys
 			} else {
-				toProvide = append(toProvide, blk.Key())
+				toProvide = append(toProvide, blkey)
 			}
 		case keysOut <- nextKey:
 			if len(toProvide) > 0 {

